### PR TITLE
Fixes #16928: Add range to enum expressions

### PR DESCRIPTION
**Pull request by peckpeck**
https://issues.rudder.io/issues/16928

### CODE REVIEW

**Comment by peckpeck** [comment]
seems ok, but expect another commit to fix tests and wanrnings

**Comment by peckpeck** [comment]
PR rebased

File: rudder-lang/src/ast/enums.rs [hunk 1]
```diff
@@ -306,31 +363,59 @@ impl<'src> EnumList<'src> {
...
+                    },
+                };
+                // get var real name
+                let varname = match var {
+                    Some(v) => v,
+                    None => treename, // TODO forbidden if tree is not global
+                                      // TODO  and some other error cases (see original version)
+                };
+                // check tat left and right are siblings
+                if let (Some(item1),Some(item2)) = (left,right) {
```
Comment (amousset) [inline rudder-lang/src/ast/enums.rs]: missing something there?

**Comment by peckpeck** [comment]
PR updated with a new commit
A cleanup commit is still missing

**Comment by peckpeck** [comment]
PR updated with a new commit
Cleanup done, unit tests OK, next: integration tests

**Comment by peckpeck** [comment]
PR updated with a new commit
Everything seems OK, all tests pass

**Comment by peckpeck** [comment]
PR updated with a new commit
Everything seems OK, all tests pass NOW

File: rudder-lang/src/ast/enums.rs [hunk 2]
```diff
@@ -6,115 +6,164 @@ use super::resource::Statement;
...
+    /// Return true if item is in the range between first and last (inclusive) None means last sibling
+    fn is_in_range(&self, item: &EnumItem<'src>, first: &Option<EnumItem<'src>>, last: &Option<EnumItem<'src>>) -> bool {
+        // 3 cases : item is a sibling, item is a descendant, item is womewhere else
+
+        // find siblings
+        let item_list = if let Some(i) = first {
+            &self.children[&self.parents[&i]]
+        } else if let Some(i) = last {
+            &self.children[&self.parents[&i]]
+        } else { panic!("Empt range") }; // else None,None is imposible
```
Comment (amousset) [inline rudder-lang/src/ast/enums.rs]: Empty

**Comment by amousset** [comment]
Can't we remove `ron` from build-deps ?

File: rudder-lang/src/ast/enums.rs [hunk 3]
```diff
@@ -6,115 +6,164 @@ use super::resource::Statement;
 use crate::error::*;
 use crate::parser::{PEnum, PEnumExpression, PSubEnum, Token};
 use std::collections::{HashMap, HashSet};
+use std::fmt;
+
+// TODO enum alias
+// TODO 
+// - trouver les todo
+
+/// This item type is internal, because First and Fast cannot be constructed from an enum declaration or from and enum expression
```
Comment (gpoblon) [inline rudder-lang/src/ast/enums.rs]: First and Last* and from an* enum expression

File: rudder-lang/src/ast/enums.rs [hunk 4]
```diff
@@ -6,115 +6,164 @@ use super::resource::Statement;
...
+            items.into_iter().map(EnumItem::Item).collect()   
+        }
+}
 
 // TODO Should a non global enum item be permitted as a variable name and must it be unique ?
 /// An enum tree is a structure containing the whole definition of an enum
 #[derive(Debug)]
-pub struct EnumTree<'src> {
+struct EnumTree<'src> {
     // a global tree has an automatic variable of its name and its items cannot be used as a variable name
```
Comment (gpoblon) [inline rudder-lang/src/ast/enums.rs]: Comment is hard to understand (and to explain), maybe something like:
The global tree name is considered a declared (and global ?) variable name. The same goes for all of its items.

File: rudder-lang/src/ast/enums.rs [hunk 5]
```diff
@@ -6,115 +6,164 @@ use super::resource::Statement;
...
-                },
-                Some(last) => {
-                    let siblings = &self.children[&self.parents[&first]];
-                    let pos1 = siblings.iter().position(|i| *i==first).unwrap();
-                    let pos2 = siblings.iter().position(|i| *i==last).unwrap();
-                    &siblings[pos1..pos2+1]
-                }
+    /// Return true if item is in the range between first and last (inclusive) None means last sibling
+    fn is_in_range(&self, item: &EnumItem<'src>, first: &Option<EnumItem<'src>>, last: &Option<EnumItem<'src>>) -> bool {
+        // 3 cases : item is a sibling, item is a descendant, item is womewhere else
```
Comment (gpoblon) [inline rudder-lang/src/ast/enums.rs]: somewhere*

File: rudder-lang/src/ast/enums.rs [hunk 6]
```diff
@@ -6,115 +6,164 @@ use super::resource::Statement;
...
+        item_position >= first_position && item_position <= last_position
+    }
+
+    /// return true if left and right have same parent
+    fn are_siblings(&self, left: Token<'src>, right: Token<'src>) -> bool {
+        self.parents.get(&EnumItem::Item(left)) == self.parents.get(&EnumItem::Item(right))
+    }
 
     /// Given a list of nodes, find the subtree that includes all those nodes and their siblings
     /// This subtree is the minimal enum tree use by an expression
```
Comment (gpoblon) [inline rudder-lang/src/ast/enums.rs]: used, or that can be used

File: rudder-lang/src/ast/enums.rs [hunk 7]
```diff
@@ -204,25 +263,19 @@ impl<'src> EnumList<'src> {
...
             self.elements.insert(*i, e.name);
         }
-        let tree = EnumTree::new(e.name, e.items, incomplete, e.global);
+        let tree = EnumTree::new(e.name, e.items, e.global);
         self.enums.insert(e.name, tree);
         Ok(())
     }
 
-    // TODO Update when parser will be modified
     /// Extend an existing enum with "enum in" from the parser
```
Comment (gpoblon) [inline rudder-lang/src/ast/enums.rs]: Aren't you talking about the "items in" ?

File: rudder-lang/src/ast/enums.rs [hunk 8]
```diff
@@ -384,45 +494,47 @@ impl<'src> EnumList<'src> {
...
     context
         .iter()
         .map(|(k, v)| format!("{}=~{}", k, v))
         .collect::<Vec<String>>()
         .join(" & ")
 }
 
+/// This is the basis of the cross iterator, it iterates over a single list
+/// of items in an enum
 // we need an item reference because ve we a pointer to the item list
```
Comment (gpoblon) [inline rudder-lang/src/ast/enums.rs]: we owe ?

File: rudder-lang/src/ast/enums.rs [hunk 9]
```diff
@@ -384,45 +494,47 @@ impl<'src> EnumList<'src> {
...
 // we cannot store the list in the same structure as its iterator
 // so let the caller store it and juste take reference
```
Comment (gpoblon) [inline rudder-lang/src/ast/enums.rs]: just*

**Comment by peckpeck** [comment]
"And maybe explain the reason why we need to pass the ".." Token" << there is already a comment in parser.rs in the structure description

File: rudder-lang/src/ast/enums.rs (hunk 4 above)
Comment (peckpeck) [reply rudder-lang/src/ast/enums.rs]: except that your wording is false
items are not variable names

**Comment by peckpeck** [comment]
PR updated with a new commit

File: rudder-lang/Cargo.toml [hunk 10]
```diff
@@ -32,6 +32,5 @@ proptest = "0.9"
 test-case = "0.3"
 
 [build-dependencies]
-ron = "0.5"
 serde = { version = "1", features = ["derive"] }
```
Comment (amousset) [inline rudder-lang/Cargo.toml]: do we still have build-dependencies actually?

Comment (gpoblon) [reply rudder-lang/Cargo.toml]: no but i'll remove it from my pr, more related

**Comment by peckpeck** [comment]
OK, squash merging this PR

**Pull request merged**